### DISCUSSION TITLE
Google photos 相册问题

### DISCUSSION THREAD

**Discussion by anwen-anyi** [AlistGo/alist#3264] [Q&A] [answered]
使用 ：https://github.com/alist-org/alist/issues/1925#issuecomment-1272334862 提到的办法拿到了code
然后 在 https://tool.nn.ci/google/request 也拿到了刷新令牌，填进去后 提示

```bash
failed get objs: failed to list objs: Request had insufficient authentication scopes.: []
```

请问是哪里问题呢~

**Comment by anwen-anyi** [comment]
@LittleJake  麻烦大佬啦QAQ

**Comment by anwen-anyi** [comment, marked as answer]
呃呃呃...解决了  还得开启 `Photos Library API` 这个API.....

**Reply by LittleJake** [reply]
对的，官方提供的令牌获取没有获取这个权限

**Reply by anwen-anyi** [reply]
> 对的，官方提供的令牌获取没有获取这个权限

下午的时候摸索出来啦，有两种情况一个是开启这个API，一个是获取Code代码的时候得勾选哪个 允许才行 两者缺一不可..

**Comment by LittleJake** [comment]
可以参考 #1925

**Reply by anwen-anyi** [reply]
谷歌相册是不是没办法 挂载单个目录（文件夹？）我尝试了相册ID或者相册的文字都不行... 似乎只能用root

临时的pr还未合并凑合看吧
https://deploy-preview-130--alist-doc.netlify.app/zh/guide/drivers/googlephotos.html

**Reply by LittleJake** [reply]
可以是可以，但是api不是同一个，后期可以添加一下

**Reply by anwen-anyi** [reply]
> 可以是可以，但是api不是同一个，后期可以添加一下

好的，麻烦大佬~ 那暂时就先这样啦

**Reply by LittleJake** [reply]
而且还有一个问题就是谷歌相册api局限性比较大，比如加载影集只能进行完全加载，不适合总是定期全量缓存。

**Reply by anwen-anyi** [reply]
刚刚说的不是很懂了（我个人是不怎么常用的....哈哈）
我看谷歌相册和网盘都是强制本地代理模式的，没有302模式
实际上我正常也是不会挂载谷歌这个的...毕竟需要proxy 我也没有国外机器...现在用的是本机用proxy接管做的教程，主要是给想挂但是不会挂的人看 具体有什么缺陷或者其他问题 不是咱们考虑的了 哈哈哈哈哈哈~~~

**Reply by LittleJake** [reply]
我这边看了一下，单独相册的话需要获取albumid，填入根文件夹ID那里。
但是目前未知如何通过**容易**的方式获取到albumid的方式。
![image](https://user-images.githubusercontent.com/13583702/216824416-778aabea-e887-4e8e-bfa4-994dadb026f7.png)
![image](https://user-images.githubusercontent.com/13583702/216824464-2bccb0c8-ccde-47b0-804c-4741bc238cd6.png)

**Reply by anwen-anyi** [reply]
比如我有一个 影集（相册）叫测试 连接如下
https://photos.google.com/album/AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K

把AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K这一串填到根文件ID哪里吗？不是这样吧我记得QAQ

你的albumID怎么获取的呢 哈哈~

**Reply by LittleJake** [reply]
我的albumid是通过[谷歌相册的API](https://developers.google.com/photos/library/guides/list#listing-albums)获取的，我测试在网页端没能抓到任何关于albumid的东西。

估计只能通过API获取了。:rofl:

**Reply by LittleJake** [reply]
> 

目测还得写个类似获取令牌的小工具来获取albumid了

**Reply by anwen-anyi** [reply]
估计是的  跟一刻相册一样了应该（这个简单登录了点一个接口链接就可以拿到）
挺麻烦的😂😂可以写上  
会不会用到时候就不知道了

**Reply by LittleJake** [reply]
> 比如我有一个 影集（相册）叫测试 连接如下 https://photos.google.com/album/AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K
> 
> 把AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K这一串填到根文件ID哪里吗？不是这样吧我记得QAQ
> 
> 你的albumID怎么获取的呢 哈哈~

[已更新](https://github.com/alist-org/docs/pull/154)

**Reply by anwen-anyi** [reply]
> > 比如我有一个 影集（相册）叫测试 连接如下 https://photos.google.com/album/AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K
> > 把AF1QipMfdFq8EpJTEPFlzwTZtCokMNe2X1ZPX54igq5K这一串填到根文件ID哪里吗？不是这样吧我记得QAQ
> > 你的albumID怎么获取的呢 哈哈~
> 
> [已更新](https://github.com/alist-org/docs/pull/154)

辛苦啦我试试看~

**Reply by anwen-anyi** [reply]
我看哪个工具还会生成一个token？相册用 Option 2 access_token的token还是用之前的token 呢？
似乎两个都行~

**Reply by LittleJake** [reply]
`client_id, client_secret, refresh_token` 或者 `access_token`，本质上还是生成access_token。

方便工具获取到refresh_token的用户，不用access_token

**Reply by anwen-anyi** [reply]
反正这俩都能用（是这样吧）...已经画好图了 哈哈...到时候说明一下就行
还有就是你看下 https://github.com/alist-org/alist/discussions/3264#discussioncomment-5022330 为啥你的这个是已经默认 勾选了 用的也是文档（开发者）提供的工具么~ 我看你截图提示的 是rclone~ 我提示的是 nn.ci

**Comment by anwen-anyi** [comment]
能麻烦大佬瞅瞅怎么回事嘛 ~ @Xhofe @LittleJake 
新问题，更换了获取Google token方式
之前是先获取哪个code，然后用code去获取刷新令牌
现在的是不用Code 直接获取刷新令牌，然后之前版本的方式不能用了...
（因为用code获取的时候需要 勾选哪个选项 下图中的，
若不勾选直接提示 `failed get objs: failed to list objs: Request had insufficient authentication scopes.: []`）
![](https://alist.nn.ci/img/drivers/google/Google-photos2.png)

**Reply by anwen-anyi** [reply]
但是我看到 https://github.com/alist-org/alist/issues/1925#issuecomment-1272334862 大佬获取的时候也没勾选 默认给你选择了似乎是~是哪里设置的问题吗

**Reply by LittleJake** [reply]
目前的情况是[谷歌废弃了OOB（Out-Of-Band）](https://developers.google.com/identity/protocols/oauth2/resources/oob-migration?hl=zh-cn)的方式获取token，存量已获取的用户不受影响，无法新增获取。
因此以前的教程没有太大参考价值，我有空会重新试一次看看如何获取。

**Reply by anwen-anyi** [reply]
> 目前的情况是[谷歌废弃了OOB（Out-Of-Band）](https://developers.google.com/identity/protocols/oauth2/resources/oob-migration?hl=zh-cn)的方式获取token，存量已获取的用户不受影响，无法新增获取。 因此以前的教程没有太大参考价值，我有空会重新试一次看看如何获取。

Yes，是谷歌把之前那种用Code获取的方式废弃了嘛~

**Reply by LittleJake** [reply]
由于OOB废弃后，需要回调url，估计要使用rclone（非中国大陆网络）来协助获取对应token。

**Reply by LittleJake** [reply]
> > 目前的情况是[谷歌废弃了OOB（Out-Of-Band）](https://developers.google.com/identity/protocols/oauth2/resources/oob-migration?hl=zh-cn)的方式获取token，存量已获取的用户不受影响，无法新增获取。 因此以前的教程没有太大参考价值，我有空会重新试一次看看如何获取。
> 
> Yes，是谷歌把之前那种用Code获取的方式废弃了嘛~

对的

**Reply by anwen-anyi** [reply]
OK 等大佬有空 瞅瞅 哈哈~

**Reply by LittleJake** [reply]
> OK 等大佬有空 瞅瞅 哈哈~

尝试过alist的新工具了吗？[Get Google Refresh Token](https://alist.nn.ci/tool/google/request.html)

**Reply by anwen-anyi** [reply]
> > OK 等大佬有空 瞅瞅 哈哈~
> 
> 尝试过alist的新工具了吗？[Get Google Refresh Token](https://alist.nn.ci/tool/google/request.html)

这个只能获取Google 网盘的token不能拿相册的

**Reply by anwen-anyi** [reply]
> > > OK 等大佬有空 瞅瞅 哈哈~
> > 
> > 
> > 尝试过alist的新工具了吗？[Get Google Refresh Token](https://alist.nn.ci/tool/google/request.html)
> 
> 这个只能获取Google 网盘的token不能拿相册的，我不知道网盘和相册是不是一样，主要是我上面标记的哪个截图里面的选项哪个选项不能勾选 是打不开的~

**Reply by xhofe** [reply]
fixed

**Reply by anwen-anyi** [reply]
> fixed

啊？啥意思QAQ~ 不是太懂....

**Reply by LittleJake** [reply]
建议再试试，我这边测试新工具是同时获取网盘和相册的权限
![image](https://user-images.githubusercontent.com/13583702/219932563-5c084b19-73be-44cb-a4d3-76cc18824aa0.png)

**Reply by LittleJake** [reply]
> fixed

你也太快了吧，哈哈:satisfied:

**Reply by anwen-anyi** [reply]
> 建议再试试，我这边测试新工具是同时获取网盘和相册的权限 

为啥我的和你的显示不一样？是俺哪里没授权嘛
下图左侧的哪个是之前获取的时候，右侧的是现在获取提示的暂时只有云盘，**`Photos Library API`** 也是开着的
![](https://img-blog.csdnimg.cn/2851bbcc8090459e86915b4d249812e6.png)

**Reply by LittleJake** [reply]
> > 建议再试试，我这边测试新工具是同时获取网盘和相册的权限
> 
> 为啥我的和你的显示不一样？是俺哪里没授权嘛 下图左侧的哪个是之前获取的时候，右侧的是现在获取提示的暂时只有云盘，**`Photos Library API`** 也是开着的 ![](https://camo.githubusercontent.com/9565258e47fc02ddee3a65512c1bc71a51dbf8d0734e275e98fabf6117ed0e43/68747470733a2f2f696d672d626c6f672e6373646e696d672e636e2f32383531626263633830393034353965383639313562346432343938313265362e706e67)

你这个是旧版本工具吧，建议重新用新的测试一次。

**Reply by anwen-anyi** [reply]
还有就是你用的是  crlone吗？ 
我用的就是[文档](https://alist.nn.ci/tool/google/request)提供的怎么是提示 开发者提供的域名 要访问呢 哈哈😂

**Reply by anwen-anyi** [reply]
~奇怪？为什么 提这个问题的时候和你没有回复的时候还没改QAQ ~
好嘛~ 34分钟前新加的 哈哈

add google photo scope
Xhofe committed 34 minutes ago
dd644b0

**Comment by anwen-anyi** [comment]
好嘛 又双叒叕 有新的问题了 似乎 是工具的问题现在 😂😂😂😂😂
提示`The refresh_token is only returned once if you use a web client.`
有视频作证 我还没拿到token 也没刷新网页..
不管勾选一个还是两个都是一样的结果QAQ
https://www.aliyundrive.com/s/ACmpLjUhm8S

**Reply by xhofe** [reply]
fixed

**Reply by anwen-anyi** [reply]
没问题了..

**Reply by anwen-anyi** [reply]
最后一个问题，可以强制默认勾选状态么？😂
 我看开始的时候云盘就是默认强制勾选状态的，总会有人粗心大意不勾选说报错的...
我看上面大佬 的就是默认允许状态了似乎~

**Reply by xhofe** [reply]
不行 这是由Google控制的 可能是需要发布/桌面应用才可以。

**Reply by anwen-anyi** [reply]
好吧（单独只有云盘的时候是默认勾选了的）~
桌面应用没有回调参数选项，等问问上面大佬发布的是啥应用
